File: bindings/rex4j/src/main/java/com/rex4j/example/RexEngine.java
package com.rex4j.example;

import com.rex4j.*;
import java.nio.ByteBuffer;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;
import org.HdrHistogram.Histogram;
import org.HdrHistogram.Recorder;
import picocli.CommandLine;
import picocli.CommandLine.Option;

public class RexEngine implements Runnable {
  private static final int TIMESTAMP_LENGTH = 8;

  @Option(
      names = {"-h", "--host"},
      required = true,
      description = "ip地址")
  String host;

  @Option(
      names = {"-p", "--port"},
      required = true,
      description = "端口")
  int port;

  @Option(
      names = {"-t", "--title"},
      defaultValue = "",
      description = "多个title用;隔开")
  String title;

  @Option(
      names = {"-y", "--type"},
      required = true,
      description = "rcv or snd")
  String type;

  @Option(
      names = {"-s", "--size"},
      defaultValue = "1024",
      description = "包大小")
  int size;

  @Option(
      names = {"-T", "--time"},
      defaultValue = "60",
      description = "发送持续时间，单位秒")
  int time;

  @Option(
      names = {"-i", "--interval"},
      defaultValue = "50",
      description = "发送间隔，单位微秒")
  long interval;

  @Option(
      names = {"-c", "--command"},
      defaultValue = "Title")
  RexCommand command;

  public static void main(String[] args) {
    System.exit(new CommandLine(new RexEngine()).execute(args));
  }

  @Override
  public void run() {
    if ("rcv".equalsIgnoreCase(type)) {
      rcv();
    } else if ("snd".equalsIgnoreCase(type)) {
      snd();
    } else {
      System.out.println("Invalid type value");
    }
  }

  void rcv() {
    AtomicLong rcv_count = new AtomicLong(0);
    Recorder recorder = new Recorder(3);

    ScheduledExecutorService statsScheduler = Executors.newSingleThreadScheduledExecutor();
    statsScheduler.scheduleAtFixedRate(
        () -> {
          Histogram hist = recorder.getIntervalHistogram();
          long count = hist.getTotalCount();
          if (count > 0) {
            double p50 = hist.getValueAtPercentile(50) / 1_000.0;
            double p90 = hist.getValueAtPercentile(90) / 1_000.0;
            double p95 = hist.getValueAtPercentile(95) / 1_000.0;
            double p98 = hist.getValueAtPercentile(98) / 1_000.0;
            double p99 = hist.getValueAtPercentile(99) / 1_000.0;
            double max = hist.getMaxValue() / 1_000.0;
            double min = hist.getMinValue() / 1_000.0;
            double mean = hist.getMean() / 1_000.0;

            System.out.printf(
                "tps: %d, mean: %.3f µs, min: %.0f µs, P50: %.0f µs, P90: %.0f µs, P95: %.0f µs,"
                    + " P98: %.0f µs, P99: %.0f µs, max: %.0f µs%n",
                count, mean, min, p50, p90, p95, p98, p99, max);
          }
        },
        1,
        1,
        TimeUnit.SECONDS);
    RexConfig config = RexConfig.builder(host, port, title).build();
    RexClient client =
        new RexClient(
            config,
            new RexHandler() {
              @Override
              public void onLogin(RexClient client, RexData data) {
                System.out.println("rcv client login ok");
              }

              @Override
              public void onMessage(RexClient client, RexData data) {
                long now = System.nanoTime();
                long timestamp = timestamp(data.getData());

                if (timestamp == -10086) {
                  System.out.printf("receive total: [%d]%n", rcv_count.get());
                  System.exit(0);
                  return;
                }
                rcv_count.incrementAndGet();
                long latency = now - timestamp;
                if (latency < 0) {
                  System.err.printf("timestamp: %d, now: %d%n", timestamp, now);
                }

                recorder.recordValue(latency);
              }
            });

    while (true) {
      try {
        Thread.sleep(1000);
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
    }
  }

  void snd() {
    RexConfig config = RexConfig.builder(host, port, title).build();
    RexClient client =
        new RexClient(
            config,
            new RexHandler() {
              @Override
              public void onLogin(RexClient client, RexData data) {
                System.out.println("snd client login ok");
              }

              @Override
              public void onMessage(RexClient client, RexData data) {
                System.out.println("snd receive: " + data.getCommand());
              }
            });

    long end = System.currentTimeMillis() + time * 1000L;
    long intervalNanos = interval * 1_000L;
    long start = 0;
    long snd_count = 0;

    do {
      start = System.nanoTime();

      RexData data = RexData.builder(command).title(title).data(wrapData(size)).build();

      client.send(data);
      snd_count += 1;

      while (true) {
        long elapsed = System.nanoTime() - start;
        if (elapsed > intervalNanos) {
          break;
        }
      }

    } while (System.currentTimeMillis() < end);

    ByteBuffer buffer = ByteBuffer.wrap(new byte[TIMESTAMP_LENGTH]);
    buffer.putLong(-10086);
    RexData stopData = RexData.builder(command).title(title).data(buffer.array()).build();
    client.send(stopData);

    System.out.printf("send total: [%d]%n", snd_count);

    client.close();
  }

  byte[] wrapData(int size) {
    ByteBuffer buffer = ByteBuffer.wrap(new byte[size]);
    long timestamp = System.nanoTime();
    buffer.putLong(timestamp);
    return buffer.array();
  }

  long timestamp(byte[] data) {
    ByteBuffer buffer = ByteBuffer.wrap(data, 0, TIMESTAMP_LENGTH);
    return buffer.getLong();
  }
}

File: bindings/rex4p/examples/rex_engine.py
#!/usr/bin/env python3
import argparse
import gc
import struct
import sys
import time
from threading import Lock, Thread

from rex4p import ClientConfig, RexClient, RexCommand, RexData, Protocol


class LatencyStats:
    """延迟统计 - GC优化版本"""

    __slots__ = ("latencies", "lock")

    def __init__(self):
        self.latencies = []
        self.lock = Lock()

    def record(self, latency_ns):
        with self.lock:
            self.latencies.append(latency_ns)

    def get_and_clear(self):
        with self.lock:
            data = self.latencies
            self.latencies = []
            return data

    def calculate_stats(self, latencies):
        if not latencies:
            return None

        count = len(latencies)
        latencies.sort()
        sorted_lat = latencies  # 重命名避免复制

        def percentile(p):
            k = (count - 1) * p / 100
            f = int(k)
            c = f + 1 if f + 1 < count else f
            return sorted_lat[f] + (k - f) * (sorted_lat[c] - sorted_lat[f])

        # 只在最后计算时转换一次
        inv_count = 1.0 / count
        min_lat = sorted_lat[0]
        max_lat = sorted_lat[-1]
        total = sum(sorted_lat)

        return {
            "count": count,
            "min": min_lat / 1000,
            "max": max_lat / 1000,
            "mean": (total * inv_count) / 1000,
            "p50": percentile(50) / 1000,
            "p90": percentile(90) / 1000,
            "p95": percentile(95) / 1000,
            "p98": percentile(98) / 1000,
            "p99": percentile(99) / 1000,
        }


class RcvHandler:
    """接收端处理器 - GC优化版本"""

    __slots__ = (
        "rcv_count",
        "stats",
        "running",
        "client",
        "should_exit",
        "_struct_unpack",
        "_stop_timestamp",
        "_zero_timestamp",
    )

    def __init__(self, client=None):
        self.rcv_count = 0
        self.stats = LatencyStats()
        self.running = True
        self.client = client
        self.should_exit = False
        # 预计算常量
        self._struct_unpack = struct.Struct("<q").unpack
        self._stop_timestamp = -10086
        self._zero_timestamp = 0
        self.start_stats_thread()

    def start_stats_thread(self):
        """启动统计线程"""

        def stats_loop():
            while self.running:
                time.sleep(1)
                latencies = self.stats.get_and_clear()
                stats = self.stats.calculate_stats(latencies)

                if stats:
                    print(
                        f"tps: {stats['count']}, "
                        f"mean: {stats['mean']:.3f} µs, "
                        f"min: {stats['min']:.0f} µs, "
                        f"P50: {stats['p50']:.0f} µs, "
                        f"P90: {stats['p90']:.0f} µs, "
                        f"P95: {stats['p95']:.0f} µs, "
                        f"P98: {stats['p98']:.0f} µs, "
                        f"P99: {stats['p99']:.0f} µs, "
                        f"max: {stats['max']:.0f} µs"
                    )

        thread = Thread(target=stats_loop, daemon=True)
        thread.start()

    def on_login(self, data):
        print("rcv client login ok")

    def on_message(self, data):
        now = time.time_ns()

        # 解析时间戳 - 使用memoryview避免复制
        data_bytes = data.data
        if len(data_bytes) >= 8:
            timestamp = self._struct_unpack(data_bytes[:8])[0]

            # 检查是否是停止信号
            if timestamp == self._stop_timestamp:
                time.sleep(0.5)
                print(f"\nreceive total: [{self.rcv_count}]")
                self.running = False
                self.should_exit = True
                return

            self.rcv_count += 1
            latency = now - timestamp

            if latency < 0:
                print(
                    f"Warning: negative latency - timestamp: {timestamp}, now: {now}",
                    file=sys.stderr,
                )
            else:
                self.stats.record(latency)


class SndHandler:
    """发送端处理器"""

    __slots__ = ()

    def on_login(self, data):
        print("snd client login ok")

    def on_message(self, data):
        print(f"snd receive: {data.command}")


# 全局缓冲区，用于减少GC
_buf_struct = struct.Struct("<q")
_buf_pack_into = _buf_struct.pack_into


def wrap_data(size, timestamp, out_buf):
    """打包数据，前8字节为时间戳 - 复用缓冲区版本"""
    _buf_pack_into(out_buf, 0, timestamp)
    return bytes(out_buf[:size])


def rcv_mode(args):
    """接收模式 - GC优化版本"""
    gc.disable()  # 禁用自动GC，由程序控制
    handler = RcvHandler()
    config = ClientConfig(
        f"{args.host}:{args.port}", Protocol.tcp(), args.title, handler
    )

    client = RexClient()
    handler.client = client  # 设置client引用
    client.connect(config)

    # 等待连接
    while not client.is_connected():
        time.sleep(0.1)

    print(f"Connected to {args.host}:{args.port}, title: {args.title}")

    # 保持运行
    try:
        while not handler.should_exit:
            time.sleep(0.1)
        print("Received stop signal, exiting...")
        client.close()
        time.sleep(0.5)  # 等待关闭完成
    except KeyboardInterrupt:
        print("\nShutting down...")
        client.close()


def snd_mode(args):
    """发送模式 - GC优化版本"""
    gc.disable()  # 禁用自动GC，由程序控制
    handler = SndHandler()
    config = ClientConfig(f"{args.host}:{args.port}", Protocol.tcp(), "", handler)

    client = RexClient()
    client.connect(config)

    # 等待连接
    while not client.is_connected():
        time.sleep(0.1)

    print(f"Connected to {args.host}:{args.port}, title: {args.title}")
    print(f"Starting to send {args.size} bytes packets for {args.time} seconds...")
    print(f"Interval: {args.interval} µs")

    # 转换命令
    command_map = {
        "Title": RexCommand.Title,
        "Group": RexCommand.Group,
        "Cast": RexCommand.Cast,
    }
    command = command_map.get(args.command, RexCommand.Title)

    end_time = time.time() + args.time
    interval_ns = args.interval * 1000  # 转换为纳秒
    snd_count = 0

    # 预分配缓冲区
    buf_size = max(args.size, 8)
    out_buf = bytearray(buf_size)
    _stop_ts = -10086

    try:
        while time.time() < end_time:
            start = time.time_ns()

            # 发送数据 - 复用缓冲区
            timestamp = time.time_ns()
            data_bytes = wrap_data(args.size, timestamp, out_buf)
            data = RexData(command, args.title, data_bytes)
            client.send(data)
            snd_count += 1

            # 精确控制发送间隔
            while True:
                elapsed = time.time_ns() - start
                if elapsed >= interval_ns:
                    break

        # 发送停止信号
        print("\nSending stop signal...")
        struct.pack_into("<q", out_buf, 0, _stop_ts)
        stop_msg = RexData(command, args.title, bytes(out_buf[:8]))
        client.send(stop_msg)

        print(f"send total: [{snd_count}]")

        time.sleep(1)  # 等待停止信号发送
        client.close()

    except KeyboardInterrupt:
        print(f"\nInterrupted. send total: [{snd_count}]")
        client.close()


def main():
    parser = argparse.ArgumentParser(description="Rex Engine Performance Test")
    parser.add_argument("-H", "--host", required=True, help="服务器IP地址")
    parser.add_argument("-p", "--port", type=int, required=True, help="服务器端口")
    parser.add_argument("-t", "--title", default="", help="标题，多个用;分隔")
    parser.add_argument(
        "-y",
        "--type",
        required=True,
        choices=["rcv", "snd"],
        help="模式：rcv(接收) 或 snd(发送)",
    )
    parser.add_argument("-s", "--size", type=int, default=1024, help="包大小(字节)")
    parser.add_argument("-T", "--time", type=int, default=60, help="发送持续时间(秒)")
    parser.add_argument("-i", "--interval", type=int, default=50, help="发送间隔(微秒)")
    parser.add_argument(
        "-c",
        "--command",
        default="Title",
        choices=["Title", "Group", "Cast"],
        help="命令类型",
    )

    args = parser.parse_args()

    if args.type == "rcv":
        rcv_mode(args)
    else:
        snd_mode(args)


if __name__ == "__main__":
    main()
